Add iterator support to url_search_params

jasnell opened the pull request:
Does two things:

* Adds simple iterators to ada::url_search_params. These are implemented as distinct types with `next()` and `is_done()` methods as opposed to traditional c++ iterators.

* Adds url search params apis to ada_c

jasnell commented:
@anonrig ... not entirely sure why the tests are failing here. They are passing locally for me :-/

anonrig commented on tests/ada_c.cpp:
```diff
@@ -258,3 +258,86 @@ TEST(ada_c, ada_get_scheme_type) {
+            "value2");
+  ASSERT_TRUE(ada_search_params_values_iter_is_done(values));
+
+  ada_string_pair pair = ada_search_params_entries_iter_next(entries);
```
You forgot to call `ada_free_string(pair.key)` for both key and value

jasnell replied:
These are `ada_string` and not `ada_owned_string`. They don't need to be freed. That'll happen when the iter is freed.

anonrig replied:
Yeah, you're right.

anonrig commented on tests/ada_c.cpp:
```diff
@@ -258,3 +258,86 @@ TEST(ada_c, ada_get_scheme_type) {
+  ada_search_params_values_iter_free(values);
+  ada_search_params_entries_iter_free(entries);
+  ada_free_owned_string(str);
+  ada_search_params_free(out);
```
Since we have `ada_free_string` and `ada_free_owned_string`, it will be better to use `ada_free_search_params` (and also for iter frees)

```suggestion
  ada_free_search_params(out);
```

anonrig commented on tests/url_search_params.cpp:
```diff
@@ -166,4 +166,40 @@ TEST(url_search_params, has) {
\ No newline at end of file
+}
+
+TEST(url_search_params, iterators) {
```
Can you also add a basic test to fuzzer (fuzz/parse.cc)?

jasnell commented on include/ada/url_search_params.h:
```diff
@@ -74,6 +91,21 @@ struct url_search_params {
+  /**
+   * @see https://url.spec.whatwg.org/#interface-urlsearchparams
+   */
+  inline url_search_params_entries_iter get_entries();
```
The `entries()`, `keys()`, and `values()` (and `forEach()` which I'm not worrying about here) are implicit in the fact that `URLSearchParams` is marked as an iterable. See https://webidl.spec.whatwg.org/#idl-iterable

anonrig replied:
I see. Can you add more descriptive comments? It would be super helpful in the long term.

anonrig commented:
@jasnell Can you add documentation to each function and a small example to README?

jasnell commented on include/ada_c.h:
```diff
@@ -109,4 +109,75 @@ const ada_url_components* ada_get_components(ada_url result);
 
+// url search params
+typedef void* ada_url_search_params;
+typedef void* ada_strings;
```
I added a comment here. Open to suggestions on the name.

anonrig reviewed (approved):
Looks good to me. I want to take a look at it after @lemire leaves a review. Thank you @jasnell

lemire commented:
@jasnell Looks fine to me. 

However, I am unsure why you don't want to use conventional C++ iterators, which allow you to automagically do this:

```C++
 for (auto key : search_params.get_keys()) {
   // do something with key
 }
```

The approach you document...

```
auto keys = search_params.get_keys();
 while (!keys.is_done()) {
   auto key = keys.next();  // "a", "c", "e", "g"
 }
```

will appear odd (I don't mean to be insulting) to many modern C++ users. It is also objectively 'busier' and easier to get wrong.

Furthermore, why 'is_done' instead of the conventional 'has_next... That is, the following is easier to understand...

```C++
auto keys = search_params.get_keys();
 while (keys.has_next()) {
   auto key = keys.next();  // "a", "c", "e", "g"
 }
```

Isn't it?


Can you elaborate on your design decisions ?

lemire commented:
Effectively, I think you need very strong arguments if you are going to bypass the standard C++ iterators. They are easy to implement, and they lead to all sorts of goodies. They are expected in C++.

jasnell commented:
Yep, I'd considered implementing normal c++ iterators but opted for this model to keep it conceptually similar to JavaScripts iterator pattern (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol), which as `next()`, `done`, etc. The key expectation being that these are most likely to be wrapped in an implementation of the `URLSearchParams` class where `next()` is called to advance.

There the pattern for wrapping these is more likely something like (forgive the mild pseudocode)

```
class URLStringParamsKeysIterator {
 public:
  struct Result {
    bool done;
    JsValue value;
  };
  Result next() {
    if (iter.done) return { true, undefined };
    return { false, iter.next() };
  }

private:
  ada::url_search_params_keys_iter iter;
}
```

I wanted also to leave room for someone to implement the conventional c++ iterator pattern if they wanted -- these should be able to co-exist. I just really have no immediate need for the c++ style iterators. I need this only to support the JS pattern.

jasnell commented:
@lemire @anonrig ... ok, I've updated to *add* the conventional c++ iterator in addition to the JS-style. For the c++ iterator I'm really just deferring to the underlying std::vector for entries, whereas the js-style has distinct instances for keys, values, and entries.

I also renamed `is_done` to `has_next` and updated accordingly. Hopefully this will strike a nice balance for both JS and C++ use.

lemire commented:
Approved.

anonrig commented on src/ada_c.cpp:
```diff
@@ -456,4 +466,225 @@ ada_owned_string ada_idna_to_ascii(const char* input, size_t length) {
+    ada_url_search_params_keys_iter result) {
+  ada::result<ada::url_search_params_keys_iter>* r =
+      (ada::result<ada::url_search_params_keys_iter>*)result;
+  if (!r) return true;
```
It seems this is more natural?

```suggestion
  if (!r) return false;
```

jasnell replied:
good catch, missed those on the update

anonrig commented on src/ada_c.cpp:
```diff
@@ -456,4 +466,225 @@ ada_owned_string ada_idna_to_ascii(const char* input, size_t length) {
+    ada_url_search_params_entries_iter result) {
+  ada::result<ada::url_search_params_entries_iter>* r =
+      (ada::result<ada::url_search_params_entries_iter>*)result;
+  delete r;
```
```suggestion
  if (r) {
    delete r;
  }
```

jasnell replied:
If the `if (r)` check should be unnecessary here, and as indicated, I'm following the existing pattern as seen in `ada_free(...)`

anonrig commented on src/ada_c.cpp:
```diff
@@ -456,4 +466,225 @@ ada_owned_string ada_idna_to_ascii(const char* input, size_t length) {
+    ada_url_search_params_entries_iter result) {
+  ada::result<ada::url_search_params_entries_iter>* r =
+      (ada::result<ada::url_search_params_entries_iter>*)result;
+  if (!r) return true;
```
```suggestion
  if (!r) return false;
```

anonrig commented on src/ada_c.cpp:
```diff
@@ -456,4 +466,225 @@ ada_owned_string ada_idna_to_ascii(const char* input, size_t length) {
+    ada_url_search_params_values_iter result) {
+  ada::result<ada::url_search_params_values_iter>* r =
+      (ada::result<ada::url_search_params_values_iter>*)result;
+  if (!r) return true;
```
```suggestion
  if (!r) return false;
```

anonrig commented on src/ada_c.cpp:
```diff
@@ -456,4 +466,225 @@ ada_owned_string ada_idna_to_ascii(const char* input, size_t length) {
+    ada_url_search_params_values_iter result) {
+  ada::result<ada::url_search_params_values_iter>* r =
+      (ada::result<ada::url_search_params_values_iter>*)result;
+  delete r;
```
```suggestion
  if (r) {
    delete r;
  }
```

jasnell replied:
Following the existing pattern here for `ada_free`. These should be fine as is.

anonrig commented on src/ada_c.cpp:
```diff
@@ -456,4 +466,225 @@ ada_owned_string ada_idna_to_ascii(const char* input, size_t length) {
+void ada_free_strings(ada_strings result) {
+  ada::result<std::vector<std::string>>* r =
+      (ada::result<std::vector<std::string>>*)result;
+  delete r;
```
Do we need if (r) check to each operation?

jasnell replied:
For frees? I don't think so. I'm following the existing pattern that is already here.

jasnell commented:
I see that many of the operations in ada_c are intended to be safe when `!r` but the current approach means that these will crash after the various `ada_free*` calls. In a separate PR I'll go through and make these all safer (not just the search_params ones) so that double frees and the other ops are more resilient after free.
